src/Emulator/Extensions/Mocks/HPSHostController.cs

```diff
@@ -171,6 +174,114 @@ private byte[] GetBytesFromSlave(int count, TimeInterval timeInterval)
+            Console.WriteLine("Starting on: "+ path);
+            currentSlave.Read(0);
+            var connection = socket.Accept();
+            Console.WriteLine("Connection established");
+
+            var commandBuffer = new byte[1];
+            int length;
+            while(SocketConnected(connection)){
+                try {
+                    PollForSocketData(connection, (int)numOfBytes.MessageType, commandBuffer);
+                }
+                catch(Exception e)
+                {
+                    Console.WriteLine("{0} Exception caught.", e);
+                    continue;
+                }
+                switch(commandBuffer[0])
+                {
+                    case (byte)SocketI2CMessageType.Stop:
+                        continue;
+                    // Parse read command
+                    case (byte)SocketI2CMessageType.Read:
+                        length = GetLength(connection);
+                    
+                        byte [] dataToSend =  GetLongBytes(length, timeInterval);
+                        try 
+                        {   
+                            connection.Send(dataToSend, 0, length, SocketFlags.None);
+                        }
+                        catch(Exception e)
+                        {
+                            Console.WriteLine("{0} Exception caught.", e);
+                            continue;
+                        }
+                        break;
+                    // Parse write command
+                    case (byte)SocketI2CMessageType.Write:
+                        length = GetLength(connection);
+                        byte[] data = new byte[length];
+                        try 
+                        {   
+                            PollForSocketData(connection, length, data);
+                            if(data.Length == 1 && data[0] == 0x86){
+                                PollForError(timeInterval);
+                            }
+                        }
+                        catch(Exception e)
+                        {
+                            Console.WriteLine("{0} Exception caught.", e);
+                            continue;
+                        }
+                        IssueCommand(data);
+                        break;
+                }
+            }
+        }
+
+        private int GetLength(Socket connection){
+            var lengthBytes = new byte[2];
+            PollForSocketData(connection, (int)numOfBytes.LengthBytes, lengthBytes);
+            int combined = lengthBytes[0] << 8 | lengthBytes[1];
+            return combined;
+        }
+
+        private void PollForSocketData(Socket connection, int length, byte[]data) {
+            if (length != data.Length){
+                throw new RecoverableException("Length to read is different from size of the buffer.");
+            }
+            int receivedLength = connection.Receive(data, 0, length, SocketFlags.None);
```

Comment:
Code here could be slightly simpler by just setting receivedLength to 0 initially